karabo_data/reader.py

```diff
@@ -662,6 +662,47 @@ def get_array(self, source, key, extra_dims=None, roi=by_index[...]):
             sorted(non_empty, key=lambda a: a.coords['trainId'][0]), dim='trainId'
         )
 
+    def get_dask_array(self, source, key, extra_dims=None):
+        import dask.array as da
+        import xarray
+        chunks = sorted(
+            self._find_data_chunks(source, key),
+            key=lambda x: x.train_ids[0] if x.train_ids.size else 0,
+        )
+
+        chunks_trainids = []
+        chunks_darrs = []
+        for chunk in chunks:
+            chunks_trainids.append(
+                self._expand_trainids(chunk.counts, chunk.train_ids)
+            )
+            chunk_dim0 = np.sum(chunk.counts)
+            chunk_shape = (chunk_dim0,) + chunk.dataset.shape[1:]
+
+            # Find chunk size of maximum 1 GB
+            # TODO: optional user control of chunking
+            limit = 1024 ** 3
+            while np.product(chunk_shape) > limit and chunk_dim0 > 1:
+                chunk_dim0 //= 2
+                chunk_shape = (chunk_dim0,) + chunk.dataset.shape[1:]
+
+            chunks_darrs.append(
+                da.from_array(chunk.dataset, chunks=chunk_shape)[chunk.slice]
+            )
+
+        darr = da.concatenate(chunks_darrs, axis=0)
```

Comment:
I am by no means an expert in dask as I for myself have just recently started using it, but couldn't you instantiate a dask.distributed client (with optional user given settings) when creating RunDir and then use it here?
I mean something like this:
```
with dask.config.set(self.dask_client.get):
       chunks_darrs.append(
                da.from_array(chunk.dataset, chunks=chunk_shape)[chunk.slice]
            )
        darr = da.concatenate(chunks_darrs, axis=0)
```
or 
``` darr = self.dask_config.sbmit(da.concatenate, chunks_arss, axis=0).result() ```

Maybe you only need to create the dask client once and all dask calls will be redirected to the client automatically, I am not sure? Regardless I think it would make sense to create in instance of a dask client which can be configured by the user (e.g. reserved computing node(s)).

Reply:
As creating a client automatically configures Dask to use it, I figured it was best not to do that in library code, but instead recommend that user code does it. This is how I've been experimenting with it:

```python
client = Client(...)
run.get_dask_array(...)
```

I thought this would allow expert users to use dask however they prefer, rather than karabo_data forcing a particular use pattern. But I'm also new to dask, so I don't know if there's a better design.

One particular doubt: with no client set up, dask defaults to using multithreading. But HDF5 & h5py don't parallelise in threads, so I think data reading is still effectively serialised. Maybe we should do something to avoid or warn about that. But the impact depends on the relative amounts of I/O and compute you want to do.